### PR TITLE
PWGEM/DQ: New functionality to check MCSignal, check if PDG is in decay hisotry

### PR DESCRIPTION
**Pull request by feisenhu**
I have added a flag that can be used selecting a prong `fPDGInHistory`.

If the variable is true, instead of performing a loop over the defined generations, 
a loop over all mothers is performed and it is checked if each provided pdg code is within the decay history.

+ some Changes to my MCtempaltes task...

### CODE REVIEW

**Review by rbailhac — COMMENTED** [review]
Dear Florian,


For the implementation in MCSignal and MCProng after the test we did, this looks good to me.

For the MCSignalLibrary, I would have small changes:



eeFromAnyCharmNoBeauty could we call it eeFromPromptCPromptC ?

eeFromAnyBeauty Could we call it eeFromAnyBAnyB ?

eeFromAnyBeautyToCharm could we call it eeFromBtoCBtoC ?



eeFromAnyBeautyAndCharm does not seem to be really what the name says. It seems to be eeFromBtoCPromptC.
eeFromAnyBeautyAndCharmBis the same here. Is it not eeFromBtoCPromptCBis ?
Both are not really the equivalent to eeFromBandBtoC. For this we would need to check in the first prompt that there is a B but no C and in the second a C and a B.

I would keep the eeFromSingleBandBtoC

Cheers,

Raphaelle

**Comment by dsekihat** [comment]
Dear @feisenhu and @rbailhac ,
Great works! For MC signals, I would ask @iarsene to review, as this may affects everywhere in DQ framework.

**Comment by iarsene** [comment]
Hi @dsekihat, thanks. I am discussing now with Florian to understand better the desired functionality

**Comment by rbailhac** [comment]
Dear @dsekihat , @iarsene , from my side it looks good.

File: PWGDQ/Core/MCSignal.h [hunk 1]
```diff
@@ -130,6 +130,80 @@ bool MCSignal::CheckProng(int i, bool checkSources, const T& track)
...
+
+    // while find mothers, check if they have provided PDG codes
+    int nIncludedPDG = 0;
+    for (int k = 1; k < fProngs[i].fPDGcodes.size(); k++) {
+      currentMCParticle = track;
+      if (!fProngs[i].fExcludePDG[k])
+        nIncludedPDG++;
+      int ith = 0;
+      while (currentMCParticle.has_mothers()) {
+        auto mother = mcStack.iteratorAt(currentMCParticle.mothersIds()[0]);
```
Comment (iarsene) [inline PWGDQ/Core/MCSignal.h:186]: there is no mcStack anymore

File: PWGDQ/Core/MCProng.h [hunk 2]
```diff
@@ -79,7 +79,7 @@ class MCProng
   MCProng();
   MCProng(int n);
   MCProng(int n, std::vector<int> pdgs, std::vector<bool> checkBothCharges, std::vector<bool> excludePDG,
-          std::vector<uint64_t> sourceBits, std::vector<uint64_t> excludeSource, std::vector<bool> useANDonSourceBitMap, bool fCheckGenerationsInTime = false);
+          std::vector<uint64_t> sourceBits, std::vector<uint64_t> excludeSource, std::vector<bool> useANDonSourceBitMap, bool fPDGInHistory = false, bool fCheckGenerationsInTime = false);
```
[the remark this answers is not in the document]
Comment (feisenhu) [reply PWGDQ/Core/MCProng.h:82]: I think there no need for it. I just checked the MCSignalLibrary. There is no signal defining `fCheckGenerationsInTime` within the prong. Normally  the setter `SetSignalInTime(true)` is used.

[the remark this answers is not in the document]
Comment (iarsene) [reply PWGDQ/Core/MCProng.h:82]: As we discussed on mattermost, independent on whether you implement a setter or not, i would like to have a constructor which initializes all the data members of the prong, so lets keep both checkGenerationsInTime and also the PDGhistory as arguments

**Review by iarsene — APPROVED** [review]

**Review by rbailhac — APPROVED** [review]

**Pull request merged**